File: gnu/gcc/libgomp/testsuite/libgomp.fortran/nestedfn3.f90
! PR middle-end/28790
! { dg-do run }

program nestomp
  integer :: j
  j = 8
  call bar
  if (j.ne.10) call abort
contains
  subroutine foo (i)
    integer :: i
  !$omp atomic
    j = j + i - 5
  end subroutine
  subroutine bar
  use omp_lib
  integer :: i
  i = 6
  call omp_set_dynamic (.false.)
  !$omp parallel num_threads (2)
    call foo(i)
  !$omp end parallel
  end subroutine
end

File: gnu/gcc/libgomp/testsuite/libgomp.c/nestedfn-3.c
/* { dg-do run } */

#include <omp.h>

extern void abort (void);

int
main (void)
{
  int i = 5, l = 0;
  int foo (void) { return i == 6; }
  int bar (void) { return i - 3; }

  omp_set_dynamic (0);

#pragma omp parallel if (foo ()) num_threads (bar ()) reduction (|:l)
  if (omp_get_num_threads () != 1)
    l = 1;

  i++;

#pragma omp parallel if (foo ()) num_threads (bar ()) reduction (|:l)
  if (omp_get_num_threads () != 3)
    l = 1;

  i++;

#pragma omp master
  if (bar () != 4)
    abort ();

#pragma omp single
  {
    if (foo ())
      abort ();
    i--;
    if (! foo ())
      abort ();
  }

  if (l)
    abort ();

  i = 8;
#pragma omp atomic
  l += bar ();

  if (l != 5)
    abort ();

  return 0;
}
